src/main/java/com/mraof/minestuck/alchemy/CardCaptchas.java

```diff
@@ -0,0 +1,274 @@
+package com.mraof.minestuck.alchemy;
+
+import com.google.common.collect.BiMap;
+import com.google.common.collect.HashBiMap;
+import com.google.common.collect.ImmutableMap;
+import com.mraof.minestuck.Minestuck;
+import com.mraof.minestuck.item.MSItems;
+import com.mraof.minestuck.network.MSPacketHandler;
+import com.mraof.minestuck.network.data.ShareCaptchasPacket;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.util.RandomSource;
+import net.minecraft.world.entity.player.Player;
+import net.minecraft.world.item.Item;
+import net.minecraftforge.event.entity.player.PlayerEvent;
+import net.minecraftforge.event.server.ServerStartedEvent;
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+import net.minecraftforge.fml.common.Mod;
+import net.minecraftforge.registries.ForgeRegistries;
+
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.*;
+
+/**
+ * Handles the creation of, and look up of, captcha codes for each registered item in the given world.
+ * The captcha is arrived at by hashing the registry name string, which gets some additional randomization via world seed.
+ * Two worlds with the same seed will produce the same captcha for a given registered item, expect in some rare instances where there was hash collision.
+ */
+@Mod.EventBusSubscriber(modid = Minestuck.MOD_ID, bus = Mod.EventBusSubscriber.Bus.FORGE)
+public class CardCaptchas
+{
+	//TODO use of the character "#" not canonically accurate and suggests there are more punch holes in a captchalogue card then there actually are
+	
+	public static final String AVAILABLE_CHARACTERS = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789!?";
+	public static final String EMPTY_CARD_CAPTCHA = "00000000";
+	
+	private static final BiMap<String, String> REGISTRY_MAP = HashBiMap.create();
```

Comment:
Using the registry name (as a string) just seemed to make sense. Eventually it also seemed to be easier to transfer when it came to packets

Reply:
For the packet part, I'd say that it is as easy as if it was resource locations or items (using `buffer.writeResourceLocation()`/`buffer.readResourceLocation()` and `buffer.writeRegistryId()`/`buffer.readRegistryId()` respectively). There is no harm in using items since this all happens after item codes have been set up. And I'd say that it is better, as it is more narrow by disallowing non-valid item registry names. Plus, after codes have been set up, I assume that the map is only used to find a code from an item object or vice versa.
(Oops, a bit longer than intended)